### ISSUE TITLE
How does "travis encrypt" work?

### ISSUE THREAD

**Issue by jameshfisher** [travis-ci/travis-ci#2982]
I can't work this out from the documentation or the `travis help encrypt` docs. What does `travis encrypt ...` actually do? Obviously, it takes some plaintext and "encrypts" it to produce some kind of cyphertext. But what kind of encryption is this? Public-key? Whose public key? Who can decrypt the value, and what do they require in order to decrypt it?

**Comment by erikdw** [comment]
@BanzaiMan : it would be awesome if some of [@jameshfisher 's wording](https://github.com/travis-ci/travis-ci/issues/2982#issuecomment-64438280) could be added into [the docs](https://docs.travis-ci.com/user/encryption-keys/).

**Comment by fallen-s4e** [comment]
Thank you. This is something important that I could not understand from the docs.

**Comment by dlenski** [comment]
Is it possible to replicate the behavior of `travis encrypt` using the `openssl` CLI?

I tried this…

```sh
echo -n 'MYPASSWORD' | openssl rsautl -encrypt -inkey MYTRAVISPUBKEY.PEM -pubin | base64 -w0
```

… and then I put the base64-encrypted output into my `.travis.yml` (as `deploy.password.secure`)

… but it just doesn't work :-(

**Comment by jameshfisher** [comment]
Three years later, added my notes to a PR https://github.com/travis-ci/docs-travis-ci-com/pull/1669

**Comment by dlenski** [comment]
I created a pure-`bash` version of `travis-encrypt`: https://github.com/dlenski/travis-encrypt.sh

Having to download 14 bajillion Ruby things in order to encrypt a frickin' password = Frustrating

**Comment by saiduttphenom** [comment]
@jameshfisher  I am working on secure code review of a project. While going through the code I have observed that in the travis.yml file there is api secure key in the file like below:

api_key:
    secure: c5qZod9MXoBTGMe2sztpvXjzgamgJi4Vm7BJSNzHcWJkG87DPgtVD (I didn't put the actual key here. It's little long)

So my question is, is it fine to publicly display this secure key in the source code. As per my assumption, it is security concern to display this key but I am not able to prove how this secure key can help the hackers in real time. To summarize, following are my questions:

1. Is it ok display the api secure key in the travis.yml publicly.
2. If should not be displayed publicly, why it shouldn't be displayed.

**Comment by dlenski** [comment]
@saiduttphenom: the "API key" is encrypted with the travis-encrypt RSA **public** key for the project. Only Travis-CI itself has the corresponding RSA **private** key.

As long as Travis-CI doesn't mishandle/leak the private key, and the RSA cryptosystem is secure… the actual plaintext value is only accessible by Travis-CI itself.

**Comment by saiduttphenom** [comment]
@dlenski : I got that. Thank you.

So the secure key displayed in the below image is a public key ???. Please refer the image:

![key](https://user-images.githubusercontent.com/31511547/43356878-5c3aba32-9296-11e8-9c97-aeaddbee4e27.png)

Sorry if I am asking anything wrong, but I want to clear it out.

**Comment by dlenski** [comment]
@saiduttphenom:

> So the secure key displayed in the below image is a public key ???

**No it is not**. As I wrote in my previous comment, it is a value that has been encrypted with an RSA public key, and it can only be decrypted with the corresponding private key, which only Travis-CI itself has.

**Comment by saiduttphenom** [comment]
@dlenski  Thanks for the response. I understood it now. 

And I got another new question about "DANGER_GITHUB_API_TOKEN". 

As told earlier I am working on secure code review of a project. While going through the travis CI logs I am able to see the “Danger_GITHUB_API_TOKEN” in the logs publicly.

But I am not able to come to a conclusion if it is a security bug because I am not sure what an external attacker can do with that token.

To summarize following are my questions:

1) Is it ok from security perspective to display these danger github API token publicly in the travis logs?
2) If it is security bug to display that, how can an external person use the token to exploit that?

Screenshot for reference:

![p3](https://user-images.githubusercontent.com/31511547/43364872-d1807d4a-9340-11e8-8cbe-147d9a857192.png)

Please see the highlighted part.

**Comment by dlenski** [comment]
@saiduttphenom, I have no idea what `DANGER_GITHUB_API_TOKEN` is, or where it came from.

Presumably it's embedded in the `travis.yml` or some other config script?
